### ISSUE TITLE
Extend tests to use an unified diff instead of a simple assert

### ISSUE THREAD

**Issue by Spacetown** [gcovr/gcovr#374]
It would be easier to check the tests if the output is an unified diff with a context ov several lines instead of a full diff.
If the test fails the clean should be skipped for easier update of the reference. The output directories should be ignored by git.

**Comment by latk** [comment]
I fully agree with all of these things. Making it easier to update the reference files is very important. 

Some context/background:

Currently HTML is just compared as a string, and the diff is produced by pytest. The diff is affected by pytest's verbosity setting: too high and the entire string is shown, too low and the diff may be abbreviated. Maybe there's a pytest plugin or something that produces a proper diff.

I'd actually prefer a proper HTML diff that ignores whitespace etc, but I haven't found a good solution. LXML has basic support for HTML diffs, but seems difficult to integrate. We currently use a similar diffing approach for XML files via the `pyutilib` library (by gcovr's original authors at Sandia) but I want to get rid of that depencency (large dependency for little value).

Feel free to change the gitignore settings, as long as the reference directories are not ignored. IIRC this could require a pattern such as:

``` .gitignore
# in gcovr/tests/.gitignore
# ignore all HTML files
*.html
# except in reference directories
!*/reference/*
```

**Comment by Spacetown** [comment]
Currently there is used following code for the check:
https://github.com/gcovr/gcovr/blob/1f17d4427c301f2a2a8bad50991f161a0e45aa5c/gcovr/tests/test_gcovr.py#L232-L233

The output comes from the assert. What about using difflib and assert that the result of the diff is "equal"?

**Comment by Spacetown** [comment]
The .gitignore works as suggested. I've also added CSS files and moved the unit test parts from the top level .gitignore to gcovr/tests/.gitignore.
I've added two options:
- `--skip_clean`: Skip the clean step after the compare (keep the output data).
- `--update_reference`: Copy the test result to the reference if they differ.

I prefer out of source builds for easier .gitignore but this will be a big change.
